hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/marker/WriteMarkers.java

```diff
@@ -52,16 +58,46 @@ public WriteMarkers(String basePath, String markerFolderPath, String instantTime
     this.instantTime = instantTime;
   }
 
+  public Option<Path> create(String partitionPath, String dataFileName, IOType type) {
+    return create(partitionPath, dataFileName, type, Option.empty());
+  }
+
   /**
    * Creates a marker without checking if the marker already exists.
    *
    * @param partitionPath partition path in the table
    * @param dataFileName data file name
    * @param type  write IO type
+   * @param handler could be empty
    * @return the marker path
    */
-  public Option<Path> create(String partitionPath, String dataFileName, IOType type) {
-    return create(partitionPath, dataFileName, type, false);
+  public Option<Path> create(String partitionPath, String dataFileName, IOType type, Option<HoodieWriteHandle> handler) {
+    boolean checkIfExists = false;
+
+    if (handler.isPresent()
+        && handler.get().getConfig().getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()
+        && handler.get().getConfig().isEarlyConflictDetectionEnable()) {
+
+      HoodieTableMetaClient metaClient = handler.get().getHoodieTableMetaClient();
+      HoodieActiveTimeline activeTimeline = metaClient.getActiveTimeline();
+
+      HoodieTimeline pendingCompactionTimeline = activeTimeline.filterPendingCompactionTimeline();
+      HoodieTimeline pendingReplaceTimeline = activeTimeline.filterPendingReplaceTimeline();
+      // TODO if current is compact or clustering then create marker directly without early conflict detection.
+      // Need to support early conflict detection between table service and common writers.
+      if (pendingCompactionTimeline.containsInstant(instantTime) || pendingReplaceTimeline.containsInstant(instantTime)) {
```

Comment:
Compaction and Clustering will create markers directly without any early conflict detection.
Bcz these hoodie table service can be async and not depend on multi-writer concurrency control.

Reply:
The details will be captured by HUDI-5809.